### DISCUSSION TITLE
AES feature

### DISCUSSION THREAD

**Discussion by blixten85** [arvidn/libtorrent#7663] [Ideas]
I request AES (military grade) encryption as an option for those who want to use encryption. Regardless if you are using a VPN or not.

The RC4 encryption is weak and thus not safe.

EDIT:
_While it is remarkable for its simplicity and speed in software, multiple vulnerabilities have been discovered in RC4, rendering it insecure._
[https://en.wikipedia.org/wiki/RC4](https://en.wikipedia.org/wiki/RC4)

ECDSA is also an option. But something that is stronger than RC4 is definitely required.

**Comment by arvidn** [comment]
RC4 is used for *obfuscation*. There is no confidentiality, as it's susceptible to man-in-the-middle attacks.
Could you elaborate what exactly you're suggesting?
In fact, https://github.com/bittorrent/bittorrent.org may be a better venue to propose extensions

**Reply by blixten85** [reply]
I want to be able to use encryption between peers (in a proper and safe way). I first suggested this at qbittorrent here on GitHub but they referred me here.

**Reply by arvidn** [reply]
In order to authenticate peers, they would need some shared secret key (or a certificate authority). There is no way to achieve this without the other peers cooperating with you, and authenticating.

To get authentication and confidentiality you can use SSL torrents.

**Reply by blixten85** [reply]
Ok, i guess i have to ask on the qbittorrent GitHub about this? 
But how do i achieve this? Why is there no tick box that just allows me to use SSl when torrenting? Certs and all fixed automatically?

This guy also asks for a more secure way to use torrents.
[https://ccextractor.org/public/gsoc/bittorrent-tls/](https://ccextractor.org/public/gsoc/bittorrent-tls/)

I assume you understand what i am asking for here, is there anything you can do to achieve this? Not just for me but for everyone in the torrent community?

**Reply by arvidn** [reply]
Here's the documentation on how to use SSL torrents: https://libtorrent.org/manual-ref.html#ssl-torrents
I would expect qBittorrent would need to add support to inject the peer certificate, if it doesn't already.

However, it sounds like this is not what you're looking for. It's a bit hard to know though, as you're quite hand-wavy in your description. It sounds like you want something with magical properties. Both confidentiality and authentication but without having to collaborate with the peers you authenticate or without a central certificate authority.

If you give up on authentication and just say; "we trust whomever we're talking to, we just don't want any 3rd party to see what we're saying.". That's problematic, what if the 3rd party man-in-the-middled our connection and that *is* the party we're talking to.
If you don't care who the other end of your connection is, it can be anybody, there is no body else left to maintain confidentiality against.

**Reply by blixten85** [reply]
I understand. It's like https in my web browser and i have to do nothing in order for this to work. I want the same for torrents. Me as end user shall not have to think about this at all. It should be ON by default. 
The changes lies with the software, i.e the torrent clients and the tracker. That's where the magic needs to happen.
And i think that the first step is the software for all users. Then trackers can follow with a high probability that end users will request this. It becomes a catch-22 situation where everyone needs to invent their own wheel and run their own tracker. (Or just shut the hell up and be glad about the neglected software we are left with to use). No offence here, but something needs to be done and it needs to be user friendly. And the ones who need to do some hand-wavy wand magic is the programmers such as yourself. Right?

**Reply by arvidn** [reply]
let's use the HTTPS analogy. In bittorrent *you* are the server. Servers accepting SSL connections (e.g. HTTPS) provide a certificate to the client, and the client authenticates it. The server operators must apply for new certificates regularly with a certificate authority. The client must trust the certificate authority that signed the server's certificate.

It sounds like you're, essentially, asking for a system with the same security properties as HTTPS, but without any of the certificate infrastructure.

> The changes lies with the software, i.e the torrent clients and the tracker. That's where the magic needs to happen.

The changes you speak of are elusive. If you think they are clear, please specify how it would work.

> And the ones who need to do some hand-wavy wand magic is the programmers such as yourself. Right?

It sounds like you're in the market for some snake oil :)

**Reply by blixten85** [reply]
No what i am asking for is that SSL functions gets fixed without me have to do anything for it. 
How this is fixed is nothing i have an opinion about, really. Except if i have to do it myself. Then i want a step by step guide on how to do it. 

I am not here for snake oil, i am here to put programmers to work. ;)

**Reply by arvidn** [reply]
the SSL functions already work, there's nothing to fix.

If you want something that doesn't provide confidentiality, but claims to; that's where the snake oil comes in.

**Reply by blixten85** [reply]
The SSL is RC4? And that is a flawed and weak encryption and it needs to be updated?

**Reply by arvidn** [reply]
RC4 is not used for SSL torrents. RC4 is only used in a separate and orthogonal feature meant to *obfuscate* the bittorrent protocol. (see the first post in this thread).

**Comment by i990049** [comment]
bitcomet support aes-128, you can use it.

**Reply by absolutep** [reply]
no, it does not. it uses same rc4 that everyone uses

**Comment by master255** [comment]
Overlay encrypted torrent networks exist, but only certain torrent clients use them.
For example, [Media Library](https://mega.nz/folder/x5FEwRoY#VcsOrnkXqCHnTdrzq0yv_g) uses Libp2p and HTTPS networks to transfer torrents.
HTTPS is a torrent web server. It requires a white IP address to work. Therefore, it is not practical. But it can also be used.
Libp2p bypasses this limitation and allows data to be transferred through two encrypted channels simultaneously. And it does not require a white IP address to work.

Therefore, in the settings, you can disable Internet access and disable UPnP publishing (this will completely disable DHT, Upnp torrents, and all Internet access). Enable torrent publishing in Libp2p and you will get torrents that have their own Libp2p DHT, run on encrypted protocols, and are completely protected from external intrusions and data decryption!

The only downside is that it only works between Media Library clients, but across the entire planet :-)
<img width="481" height="868" alt="image" src="https://github.com/user-attachments/assets/c88214ae-7f19-4bbc-b4b7-5104d6c0e60d" />

**Comment by arvidn** [comment]
uTorrent supports a feature "encrypted torrents", where the file content is encrypted as it's read from disk, and decrypted when written to disk. i.e. all hashing and network messages happen on *encrypted* data.

It uses a symmetric 256 bit AES key, which is optional. Without the key you can download and seed the content, but it will be encrypted on disk.

With the key, your files on disk will be in clear text. We made it possible to pass along the key as a magnet link parameter.

One use case we had in mind at the time was for a service like Bittorrent Sync (now Resilio Sync), where you may want 3rd parties to help you keep your content online and available, but without having access to the clear text.

**Reply by master255** [reply]
So it can be transferred as a password-protected archive. But it would be nice if your library could do this using parameters.
